Implementation of saving and loading customization parameters and previously loaded graphs

RodionovMaxim05 opened the pull request:
Добавлена реализация сохранения и загрузки параметров кастомизации и ранее загруженных графов. Также исправлен баг, связанный с перекрашиванием одной вершины при применении 'CircularPlacementStrategy'.

Данный PR решает проблему [#35](https://github.com/spbu-coding-2023/graphs-graphs-12/issues/35)

RodionovMaxim05 committed:
feat: implement saving and loading of settings parameters

RodionovMaxim05 committed:
fix: a bug in 'CircularPlacementStrategy'
Fix a bug where one vertex is repainted gray after applying 'CircularPlacementStrategy'

RodionovMaxim05 committed:
feat: implement saving and loading the history of previously opened graphs

RodionovMaxim05 committed:
docs: add doc-strings for application variable information

IlyaSuponev committed:
Merge branch 'linter' into config

IlyaSuponev commented:
Конфликты разрешены в соответствии с требованиями ветки `linter`.

IlyaSuponev commented on application/src/main/kotlin/Main.kt:
```diff
@@ -1,33 +1,156 @@
+/**
+ * Data class representing the settings for application.
+ *
+ * @property currentStyle The current style of application, represented by a MutableState of [JetStyle]
+ * @property currentFontSize The current font size of text in application, represented by a MutableState of [JetSize]
+ * @property currentCornersStyle The current corner style of application, represented by a MutableState of [JetCorners]
+ * @property currentFontFamily The current font family of text in application, represented by a MutableState of [JetFontFamily]
+ * @property isDarkMode A boolean MutableState indicating whether the application is in dark mode or not
```
У нас в проекте принято, что описание свойств начинается с маленькой буквы.

```suggestion
 * @property currentStyle the current style of application, represented by a MutableState of [JetStyle]
 * @property currentFontSize the current font size of text in application, represented by a MutableState of [JetSize]
 * @property currentCornersStyle the current corner style of application, represented by a MutableState of [JetCorners]
 * @property currentFontFamily the current font family of text in application, represented by a MutableState of [JetFontFamily]
 * @property isDarkMode a boolean MutableState indicating whether the application is in dark mode or not
```

RodionovMaxim05 replied:
Исправил

IlyaSuponev commented on application/src/main/kotlin/Main.kt:
```diff
@@ -1,33 +1,156 @@
+ * @property currentFontFamily The current font family of text in application, represented by a MutableState of [JetFontFamily]
+ * @property isDarkMode A boolean MutableState indicating whether the application is in dark mode or not
+ */
+data class JetSettings(
+	val currentStyle: MutableState<JetStyle>,
+	val currentFontSize: MutableState<JetSize>,
+	val currentCornersStyle: MutableState<JetCorners>,
+	val currentFontFamily: MutableState<JetFontFamily>,
+	val isDarkMode: MutableState<Boolean>
+)
```
Я бы вынес в отдельный файл, я это и сделаю при проверке после PR на линтере, он будет ругаться)

IlyaSuponev replied:
Это можешь не править, я сам поправлю, перед PR из `linter` в `main`.

RodionovMaxim05 replied:
Ок

IlyaSuponev commented on application/src/main/kotlin/Main.kt:
```diff
@@ -1,33 +1,156 @@
+			isDarkMode = remember { mutableStateOf(isDarkModeValue) }
+		)
+
+		val file = File("../settings")
+		findOrCreateFile(file)
+		loadSettings(file, jetSettings)
```
Я бы вынес это в класс `SettingsModel`. Чтобы при инициализации подтягивалось информация из файла настроек. 

Также возможно надо создовать файл настроек в папке `user.home` из окружения:
[See more](https://docs.oracle.com/javase/8/docs/api/java/lang/System.html#getenv-java.lang.String-).

IlyaSuponev replied:
И так норм, но как по мне делегировать классу инициализацию, будет проще.
Есть интерфейс `Closable`, который при удалении объекта вызовет метод `close()`.

RodionovMaxim05 replied:
1) Сделал. 2) Мне кажется, что лучше все-таки данные файлы хранить внутри проекта, т.к. они не несут никакой информации вне самого проекта. 3) Не получилось сделать(

RodionovMaxim05 committed:
refactor: move everything related to loading customization parameters to 'SettingsModel'
Also recover deleted lines of code after merging the "linter" branch into "config".
